orttraining/orttraining/core/optimizer/compute_optimizer/padding_elimination.cc

```diff
@@ -0,0 +1,550 @@
+        InsertNodesForInput(graph, *node, i, squeeze_out_arg, logger);
+      }
+    }
+  }
+
+  std::vector<int64_t> first_indices;
+  first_indices.push_back(0);
+  ONNX_NAMESPACE::TensorProto first_indice_const_tensor;
+  first_indice_const_tensor.set_name(graph.GenerateNodeArgName("indices"));
+  first_indice_const_tensor.set_data_type(ONNX_NAMESPACE::TensorProto_DataType_INT64);
+  first_indice_const_tensor.add_dims(first_indices.size());
+  first_indice_const_tensor.set_raw_data(first_indices.data(), first_indices.size() * sizeof(int64_t));
+  NodeArg* first_index_arg = &graph_utils::AddInitializer(graph, first_indice_const_tensor);
+
+  // Get the first dim value of flattened input_ids which is batch_size * seq_len
+  NodeArg* first_dim = GetDimsValue(graph, reshape_output_args[0], first_index_arg, *embedding_node);
+
+  std::vector<int64_t> first_two_indices{0, 1};
+  ONNX_NAMESPACE::TensorProto first_two_indices_const_tensor;
+  first_two_indices_const_tensor.set_name(graph.GenerateNodeArgName("first_two_indices"));
+  first_two_indices_const_tensor.set_data_type(ONNX_NAMESPACE::TensorProto_DataType_INT64);
+  first_two_indices_const_tensor.add_dims(first_two_indices.size());
+  first_two_indices_const_tensor.set_raw_data(first_two_indices.data(), first_two_indices.size() * sizeof(int64_t));
+  NodeArg* first_two_indices_arg = &graph_utils::AddInitializer(graph, first_two_indices_const_tensor);
+  // Get the first two dims value of input_ids which is [batch_size, seq_len]
+  NodeArg* first_two_dims_arg = GetDimsValue(graph, input_ids_arg, first_two_indices_arg, *embedding_node);
+
+  // Add pattern to each output node of the subgraph
+  // to unflatten the shape of [valid_token_count, ...] to [batch_size, seq_len, ...]
+  for (const auto& node : candidate_output) {
+    for (size_t i = 0; i < node->InputDefs().size(); ++i) {
+      if (candidate_edges.find(node->MutableInputDefs()[i]) != candidate_edges.end()) {
+        // Get a shape of the i-th input of the node with first index updated to value of first_dim
+        // which is batch_size * seq_len. This shape arg will be used as the shape input of GatherGrad
+        NodeArg* shape_arg_for_gather_grad = UpdateShape(
+            graph, node->MutableInputDefs()[i], first_dim, first_index_arg, *node);
+        InsertNodesForOutput(graph, *node, i, squeeze_out_arg, shape_arg_for_gather_grad, first_two_dims_arg, logger);
+      }
+    }
+  }
+
+  // Update shape for each edge of the subgraph
+  for (auto edge : candidate_edges) {
+    ONNX_NAMESPACE::TensorShapeProto flattened_shape;
+    flattened_shape.add_dim()->set_dim_param(token_dim_name);
+    auto input_shape = edge->Shape();
+    for (int k = 2; k < input_shape->dim_size(); k++) {
+      ORT_ENFORCE(input_shape->dim(k).has_dim_value());
+      flattened_shape.add_dim()->set_dim_value(input_shape->dim(k).dim_value());
+      edge->SetShape(flattened_shape);
+    }
+  }
```

Comment:
where did you ClearShape, and reset modified = True?

Reply:
do we still need clear shape?

Reply:
Yes, I will do it after this pr as soon as possible.